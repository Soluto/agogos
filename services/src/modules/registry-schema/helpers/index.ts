import proceedPolicy from './policy-helper';

export { proceedPolicy };
