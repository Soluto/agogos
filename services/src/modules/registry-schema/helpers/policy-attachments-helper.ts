import * as _ from 'lodash';
import * as opaHelper from '../../opa-helper';
import logger from '../../logger';
import { PolicyType, ResourceMetadata, ResourceRepository } from '../../resource-repository';
import resourceRepository from '../repository';
import { ResourceMetadataInput, PolicyInput } from '../types';

type TempLocalPolicyAttachment = { metadata: ResourceMetadataInput; path: string; type: PolicyType };

const policyMetadataComparer = (p1: PolicyInput, p2: PolicyInput) =>
  p1.metadata.namespace === p2.metadata.namespace && p1.metadata.name === p2.metadata.name;

export default class PolicyAttachmentsHelper {
  protected policyAttachmentsToSave: TempLocalPolicyAttachment[] = [];
  protected policyAttachmentsToDelete: PolicyInput[] = [];

  async sync(oldState: PolicyInput[], newState: PolicyInput[]) {
    // Policy definitions generated by plugins will always be re-generated here. This is probably
    // better than reading the previous gateway resource group just for this comparison
    const addedOrUpdatedPolicies = _.differenceWith(newState, oldState, _.isEqual);
    await this.generate(addedOrUpdatedPolicies);

    const removedPolicies = _.differenceWith(oldState, newState, policyMetadataComparer);
    this.policyAttachmentsToDelete.push(...removedPolicies);
  }

  private async generate(policies: PolicyInput[]) {
    for (const policy of policies) {
      if (!policyAttachmentStrategies[policy.type]) continue;

      const attachment = await policyAttachmentStrategies[policy.type].generate(policy);
      this.policyAttachmentsToSave.push(attachment);
    }
  }

  async writeToRepo() {
    for (const attachment of this.policyAttachmentsToSave) {
      await policyAttachmentStrategies[attachment.type].saveToRepo(resourceRepository, attachment);
    }

    for (const policy of this.policyAttachmentsToDelete) {
      await policyAttachmentStrategies[policy.type].removeFromRepo(resourceRepository, policy.metadata);
    }
  }

  async cleanup() {
    for (const attachment of this.policyAttachmentsToSave) {
      await policyAttachmentStrategies[attachment.type].cleanup(attachment);
    }
  }
}

const policyAttachmentStrategies = {
  [PolicyType.opa]: {
    generate: async (input: PolicyInput) => {
      const path = await opaHelper.prepareCompiledRegoFile(input.metadata, input.code);
      return { path, metadata: input.metadata, type: PolicyType.opa };
    },
    cleanup: async (attachment: TempLocalPolicyAttachment) => {
      try {
        await opaHelper.deleteLocalRegoFile(attachment.path);
      } catch (err) {
        logger.warn(
          { err, attachment },
          'Failed cleanup of compiled rego file, this did not affect the request outcome'
        );
      }
    },
    saveToRepo: async (resourceRepository: ResourceRepository, attachment: TempLocalPolicyAttachment) => {
      const compiledRego = await opaHelper.readLocalRegoFile(attachment.path);
      const filename = opaHelper.getCompiledFilename(attachment.metadata);
      await resourceRepository.writePolicyAttachment(filename, compiledRego);
    },
    removeFromRepo: async (resourceRepository: ResourceRepository, metadata: ResourceMetadata) => {
      const filename = opaHelper.getCompiledFilename(metadata);
      await resourceRepository.deletePolicyAttachment(filename);
    },
  },
};
