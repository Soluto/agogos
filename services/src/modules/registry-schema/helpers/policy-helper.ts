import { parseType } from 'graphql';
import { PolicyInput } from '..';
import logger from '../../logger';
import { PolicyArgDefinition } from '../../resource-repository';

export function markOptionalPolicyArgs(policies?: PolicyInput[]) {
  policies?.forEach(p => {
    if (!p.args) return;

    Object.entries(p.args).forEach(([argName, argDef]: [string, PolicyArgDefinition]) => {
      try {
        const type = parseType(argDef.type);
        argDef.optional = type.kind !== 'NonNullType' || !!argDef.default;
      } catch (err) {
        logger.warn({ policy: p.metadata, argName, argDef, err }, 'Invalid policy argument');
        throw err;
      }
    });
  });
}
