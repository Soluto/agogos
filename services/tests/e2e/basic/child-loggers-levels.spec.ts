import { startContainerOutputCapture } from '../../helpers/get-container-logs';
import { updateGatewaySchema } from '../../helpers/utility';

describe('Child loggers levels', () => {
  const validApiKey = 'validApiKey';
  const invalidApiKey = 'invalidApiKey';

  test('No logs for valid key', async () => {
    const endContainerOutputCapture = await startContainerOutputCapture('gateway');
    const resp = await updateGatewaySchema('http://localhost:8080', validApiKey);
    expect(resp.status).toEqual(200);

    const captureResult = await endContainerOutputCapture();
    expect(captureResult).toMatchSnapshot();
  });

  test('No trace logs for invalid key', async () => {
    const endContainerOutputCapture = await startContainerOutputCapture('gateway');
    const resp = await updateGatewaySchema('http://localhost:8080', invalidApiKey);
    expect(resp.status).toEqual(401);

    const captureResult = await endContainerOutputCapture();
    expect(captureResult).toMatchSnapshot();
  });
});
