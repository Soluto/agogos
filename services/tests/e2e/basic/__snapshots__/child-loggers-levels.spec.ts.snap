// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Child loggers levels No logs for valid key 1`] = `
"**** START OUTPUT CAPTURE *****\\"
\\"***** END OUTPUT CAPTURE *****"
`;

exports[`Child loggers levels No trace logs for invalid key 1`] = `
"**** START OUTPUT CAPTURE *****\\"
[hh:mm:ss.ms] DEBUG (auth-strategy-api-key): Unexpected api key
    apiKey: \\"in***ey\\"
\\"***** END OUTPUT CAPTURE *****"
`;
